merkle/log_verifier.go

```diff
@@ -58,46 +58,62 @@ func (m LogVerifier) VerifyInclusionProof(leafIndex, treeSize int64, proof [][]b
 }
 
 // RootFromInclusionProof calculates the expected tree root given the proof and leaf.
-// leafIndex starts at 0. treeSize starts at 1.
-func (m LogVerifier) RootFromInclusionProof(leafIndex, treeSize int64, proof [][]byte, leaf []byte) ([]byte, error) {
-	if leafIndex >= treeSize {
-		return nil, fmt.Errorf("leafIndex %d > treeSize %d", leafIndex, treeSize)
+// leafIndex starts at 0.  treeSize is the number of nodes in the tree.
+// proof is an array of neighbor nodes from the bottom to the root.
+func (v LogVerifier) RootFromInclusionProof(leafIndex, treeSize int64, proof [][]byte, leaf []byte) ([]byte, error) {
+	if leafIndex < 0 {
+		return nil, errors.New("invalid leafIndex < 0")
 	}
-	if leafIndex < 0 || treeSize < 1 {
-		return nil, errors.New("leafIndex < 0 or treeSize < 1")
+	if treeSize < 0 {
+		return nil, errors.New("invalid treeSize < 0")
+	}
+	lastIndex := treeSize - 1 // Rightmost node in tree.
+	if leafIndex > lastIndex {
+		return nil, fmt.Errorf("leafIndex is not in a tree of size %d, want %d<%d", treeSize, leafIndex, treeSize)
 	}
 
-	nodeIndex := leafIndex
-	lastNode := treeSize - 1
-	nodeHash := m.hasher.HashLeaf(leaf)
+	cntIndex := leafIndex
+	cntHash := v.hasher.HashLeaf(leaf)
 	proofIndex := 0
 
-	for lastNode > 0 {
-		if proofIndex == len(proof) {
+	// Tree is numbered as follows, where nodes at each level are counted from left to right.
+	//       0
+	//     0   1
+	//   0  1 2  3
+
+	// Hash sibling nodes into the current hash starting at the leaf and continuing to the root.
+	// Use the highest order 1 bit in the rightmost node as the stopping condition.
+	for lastIndex > 0 {
+		if proofIndex >= len(proof) {
 			return nil, fmt.Errorf("insuficient number of proof components (%d) for treeSize %d", len(proof), treeSize)
 		}
-		if isRightChild(nodeIndex) {
-			nodeHash = m.hasher.HashChildren(proof[proofIndex], nodeHash)
+		if isRightChild(cntIndex) {
+			cntHash = v.hasher.HashChildren(proof[proofIndex], cntHash)
 			proofIndex++
-		} else if nodeIndex < lastNode {
-			nodeHash = m.hasher.HashChildren(nodeHash, proof[proofIndex])
+		} else if cntIndex < lastIndex {
+			// Test whether a sibling node to the right exists at this level.
```

Comment:
Remove the "test whether". If the comment was outside the 'else' it would be ok.